ush/python/pygfs/task/archive.py

```diff
@@ -417,3 +438,151 @@ def replace_string_from_to_file(filename_in, filename_out, search_str, replace_s
         replace_string_from_to_file(in_track_p_file, out_track_p_file, "AVNO", pslot4)
 
         return
+
+    @staticmethod
+    @logit(logger)
+    def _archive_expdir(arch_dict: Dict[str, Any]) -> bool:
+        """
+        This function checks if the EXPDIR should be archived this RUN/cycle
+        and returns the temporary path in the ROTDIR where the EXPDIR will be
+        copied to for archiving.
+
+        Parameters
+        ----------
+        arch_dict: Dict
+            Dictionary with required parameters, including the following:
+
+            current_cycle: Datetime
+                Date of the current cycle.
+            SDATE: Datetime
+                Starting cycle date.
+            EDATE: Datetime
+                Ending cycle date.
+            NET: str
+                The workflow type (gfs or gefs)
+            ARCH_EXPDIR_FREQ: int
+                Frequency to perform EXPDIR archiving
+            ROTDIR: str
+                Full path to the ROTDIR
+        """
+
+        # Get commonly used variables
+        current_cycle = arch_dict.current_cycle
+        sdate = arch_dict.SDATE
+        edate = arch_dict.EDATE
+        mode = arch_dict.MODE
+        assim_freq = to_timedelta(f"+{arch_dict.assim_freq}H")
+        # Convert frequency to seconds from hours
+        freq = arch_dict.ARCH_EXPDIR_FREQ * 3600
+
+        # Skip gfs and enkf cycled RUNs (only archive during gdas RUNs)
+        # (do not skip forecast-only, regardless of RUN)
+        if arch_dict.NET == "gfs" and arch_dict.MODE == "cycled" and arch_dict.RUN != "gdas":
+            return False
+
+        # Determine if we should skip this cycle
+        # If the frequency is set to 0, only run on sdate (+assim_freq for cycled) and edate
+        if freq == 0:
+            if mode == "forecast-only" and (current_cycle != sdate or current_cycle != edate):
+                return False
+            elif mode == "cycled" and (current_cycle != sdate + assim_freq or current_cycle != edate):
+                return False
+        # Otherwise, the frequency is in hours
+        elif mode == "forecast-only" and (sdate - current_cycle).total_seconds() % freq != 0:
+            return False
+        elif mode == "cycled" and (sdate + assim_freq - current_cycle).total_seconds() % freq != 0:
+            return False
+
+        # Looks like we are archiving the EXPDIR
+        return True
```

Comment:
Reversing the polarity of the ifs, because I think we always want to save the first and last, even if the end time is not a multiple of the save frequency.
```suggestion
        first_full = sdate
        if mode in ["cycled"]:
            first_full += assim_freq
        if current_cycle in [first_full, edate]:
            # Always save the first and last
            return True
        elif (current_cycle - first_full).total_seconds() % freq == 0:
            # Otherwise, the frequency is in hours
            return True
        else
            return False
```

Reply:
Good catch.  Thanks.